### PR TITLE
chore: Refactored Button styles to use `createStyles` utility

### PR DESCRIPTION
**Pull request by josh-bagwell**
## Summary

Style refactor for buttons to work with new theming, branding, tokens, createStyles utility and `cs` prop. 

Fixes: #2274  

We want to more easily apply theming to Buttons and the styles need to be refactored to more easily apply those themed styles with new tokens. 

## Release Category
Components

### BREAKING CHANGES
- We refactored how we styled Buttons to use our `createStyles` utility function. We don't anticipate
this as a breaking change but, there may be slight changes to visual test.
- Icons will no longer be "filled" on toggle. This decision was made to not have the existing icon
look different in the toggled state from default state.
- `PrimaryButton`: On the `inverse` variant, the focus ring is now consistent with the default variant of `PrimaryButton`. This will visually change the `inverse` variant to have a larger appearance when focused. 
- `colors` will no longer support the `focusRing` option:
  ```tsx
  import {focusRing} from '@workday/canvas-kit-react/common';

  // before
  <PrimaryButton
    colors={{
      // other colors
      focus: {
        // other colors
      focusRing: focusRing(/* options */)
      }
    }}
  />

  // after
  <PrimaryButton
    colors={{
      // other colors
      focus: {
        // other colors
      }
    }}
    css={{
      ':focus-visible': focusRing(/* options */)
    }}
  />;
  ```
---

## For the Reviewer

- [x] PR title is short and descriptive
- [x] PR summary describes the change (Fixes/Resolves linked correctly)
- [x] Breaking Changes provides useful information to upgrade to this code or removed if not applicable

## Where Should the Reviewer Start?

`/modules/react/button`

### CODE REVIEW

File: modules/react/button/lib/BaseButton.tsx [hunk 1]
```diff
@@ -292,13 +133,295 @@ export const getPaddingStyles = (
...
+          buttonVars.active(colors?.active || {}),
+          buttonVars.default(colors?.default || {}),
+          buttonVars.disabled(colors?.disabled || {}),
+          buttonVars.focus(colors?.focus || {}),
+          buttonVars.hover(colors?.hover || {}),
+        ]}
+        {...elemProps}
+      >
         {children}
-      </ButtonContainer>
```
Comment (NicholasBoll) [inline modules/react/button/lib/BaseButton.tsx:301]: The `cs` prop should always go last. Otherwise if someone passes a `className`, all styling will be overridden.

File: modules/react/button/lib/BaseButton.tsx [hunk 2]
```diff
@@ -292,13 +133,295 @@ export const getPaddingStyles = (
...
+            size: size,
+            iconPosition: getIconPosition(size, iconPosition),
+          }),
+          buttonVars.active(colors?.active || {}),
+          buttonVars.default(colors?.default || {}),
+          buttonVars.disabled(colors?.disabled || {}),
+          buttonVars.focus(colors?.focus || {}),
+          buttonVars.hover(colors?.hover || {}),
+        ]}
+        {...elemProps}
```
Comment (NicholasBoll) [inline modules/react/button/lib/BaseButton.tsx:421]: The `cs` prop should always go last. Otherwise if someone passes a `className`, all styling will be overridden.

Comment (josh-bagwell) [reply modules/react/button/lib/BaseButton.tsx:384]: Done

**Comment by mannycarrera4** [comment]
image
 looks like icons action bar needs to be updated

**Comment by mannycarrera4** [comment]
In the PR the focus button seems like it has a white background
image

In master, it should have a grey background
image

**Comment by mannycarrera4** [comment]
In the PR toolbar icon button isn't filled:
image
What's in master is filled:
image

File: yarn.lock [hunk 3]
```diff
@@ -18131,9 +18136,9 @@ windows-release@^3.1.0:
     execa "^1.0.0"
 
 word-wrap@^1.0.3, word-wrap@~1.2.3:
-  version "1.2.4"
-  resolved "https://registry.yarnpkg.com/word-wrap/-/word-wrap-1.2.4.tgz#cb4b50ec9aca570abd1f52f33cd45b6c61739a9f"
-  integrity sha512-2V81OA4ugVo5pRo46hAoD2ivUJx8jXmWXfUkY4KFNw0hEptvN0QfH3K4nHiwzGeKl5rFKedV48QVoqYavy4YpA==
+  version "1.2.3"
```
Comment (mannycarrera4) [inline yarn.lock:18139]: did we want to update this?

File: modules/react/popup/lib/PopupCloseIcon.tsx [hunk 4]
```diff
@@ -21,7 +21,7 @@ export const PopupCloseIcon = createSubcomponent('button')({
       icon={xIcon}
       type="button"
       position="absolute"
-      right={space.xxxs}
+      insetInlineEnd={space.xxxs}
```
Comment (mannycarrera4) [inline modules/react/popup/lib/PopupCloseIcon.tsx:24]: nice!

File: modules/docs/mdx/10.0-UPGRADE_GUIDE.mdx [hunk 5]
```diff
@@ -123,6 +128,22 @@ these sizes should change along with it. Since `px` are a fixed unit and do not
...
+
+#### Button Focus Ring Update
+
+We found that focus ring was not consistent across all Buttons. We have made some changes to create
+a consistent focus ring across all Buttons and variants. These will need to be taken note of due to
+some small changes visually with Buttons.
+
+- `PrimaryButton`: `inverse` variant now has the same focus ring as default variant and as
+  `SecondaryButton`.
+
```
Comment (NicholasBoll) [inline modules/docs/mdx/10.0-UPGRADE_GUIDE.mdx:146]: We should highlight the breaking change that `colors` no longer supports the `focusRing` option.

```jsx
import {focusRing} from '@workday/canvas-kit-react/common';

// before
<PrimaryButton
  colors={{
    // other colors
    focus: {
      // other colors
     focusRing: focusRing(/* options */)
    }
  }}
/>

// after
<PrimaryButton
  colors={{
    // other colors
    focus: {
      // other colors
    }
  }}
  css={{
    ':focus-visible': focusRing(/* options */
  }}
/>
```

File: modules/preview-react/pill/lib/PillIconButton.tsx [hunk 6]
```diff
@@ -63,6 +62,11 @@ const StyledIconButton = styled(BaseButton)<StyledType & PillIconButtonProps>({
     pointerEvents: 'all',
     cursor: 'pointer',
   },
+  '&:focus, &:focus-visible': {
```
Comment (NicholasBoll) [inline modules/preview-react/pill/lib/PillIconButton.tsx:65]:
```suggestion
  '&.focus, &:focus-visible': {
```

File: modules/react/button/lib/BaseButton.tsx [hunk 7]
```diff
@@ -292,13 +126,265 @@ export const getPaddingStyles = (
...
+      height: cssVar(system.space.x6, space.m),
+      minWidth: 'auto',
+      paddingInline: cssVar(system.space.x3, space.xs),
+      gap: cssVar(system.space.x1, space.xxxs),
+    }),
+  },
+  iconPosition: {
+    largeOnly: createStyles({
+      padding: '0',
+      minWidth: cssVar(system.space.x12, '3rem'),
```
Comment (NicholasBoll) [inline modules/react/button/lib/BaseButton.tsx:295]: I'd like to get rid of all fallbacks to system variables. The optimizer will be able to parse the CSS files from our tokens package and automatically fill in fallbacks. This would be a win for us in Canvas Kit because we don't have to lookup fallbacks manually, but still get the benefit of working in environments where the variables are not provided (most likely Storybook)

File: modules/react/common/lib/CanvasProvider.tsx [hunk 8]
```diff
@@ -1,34 +1,60 @@
...
-    },
-  };
+const mappedKeys = {
+  lightest: 'lightest',
+  light: 'light',
+  main: 'base',
+  dark: 'dark',
+  darkest: 'darkest',
+  contrast: 'accent',
+};
```
Comment (mannycarrera4) [inline modules/react/common/lib/CanvasProvider.tsx:18]: is this something that consumer will have to do if they're using canvas provider?

File: .storybook/preview.js [hunk 9]
```diff
@@ -34,7 +37,9 @@ function storySort(a, b) {
     prefix('basic', 'aa'),
     prefix('default', 'ab'),
     prefix('testing', 'zzz'),
-    prefix('examples', 'zz')
+    prefix('examples', 'zz'),
+    // Make sure upgrade guides follow chronological order by replacing `v9.0` with `v09.0`
+    value => value.replace(/v([1-9]\-0)/, '0$1')
```
Comment (mannycarrera4) [inline .storybook/preview.js:42]: 👍

File: modules/preview-react/pill/lib/Pill.tsx [hunk 10]
```diff
@@ -158,6 +156,13 @@ const StyledNonInteractivePill = styled(StyledBasePill)<StyledType>({
   cursor: 'default',
   overflow: 'revert', // override BaseButton overflow styles so the click target exists outside the pill for removable
   position: 'relative',
+  '&.focus-visibile, &.focus': {
```
Comment (mannycarrera4) [inline modules/preview-react/pill/lib/Pill.tsx:159]:
```suggestion
  '&:focus-visibile, &.focus': {
```

File: modules/react/button/lib/Button.tsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            size={size}
+            iconPosition={iconPosition}
+            icon={icon}
+            shouldMirrorIcon={shouldMirrorIcon}
+          />
+        )}
+      </BaseButton>
+    );
+  },
+});
```
Comment (mannycarrera4) [inline modules/react/button/lib/Button.tsx:67]: Is this file still needed or is it replacing `BaseButon`?

**Comment by josh-bagwell** [comment]
> image looks like icons action bar needs to be updated

Should we consider making this a test as well in the future? I wouldn't have noticed this without you mentioning it to me. 

I had to reset a default `iconPosition`.

File: modules/react/popup/lib/PopupCloseIcon.tsx (hunk 4 above)
Comment (josh-bagwell) [reply modules/react/popup/lib/PopupCloseIcon.tsx:24]: This is what we'll be moving to with other components as well so we no longer have to worry about RTL.

File: modules/docs/mdx/10.0-UPGRADE_GUIDE.mdx (hunk 5 above)
Comment (josh-bagwell) [reply modules/docs/mdx/10.0-UPGRADE_GUIDE.mdx:146]: Done.

**Comment by mannycarrera4** [comment]
> > In the PR toolbar icon button isn't filled: image What's in master is filled: image
> 
> This was a change we talked with design on. We decided that the icon shouldn't changed when toggled. This came up with segmented control on the tablet icon. It looks completely different so we decided to change it and not have it fill.

We should probably add this as change to the upgrade guide

File: modules/preview-react/pill/lib/Pill.tsx [hunk 12]
```diff
@@ -147,8 +138,15 @@ const StyledBasePill = styled(BaseButton.as('button'))<StyledType & PillProps>(
     }),
   },
   ({variant}) => ({
-    '&:focus': {
+    '&:focus, &:focus-visible': {
```
Comment (mannycarrera4) [inline modules/preview-react/pill/lib/Pill.tsx:141]: does this need the `.focus` class?

File: modules/react/common/lib/CanvasProvider.tsx (hunk 8 above)
Comment (josh-bagwell) [reply modules/react/common/lib/CanvasProvider.tsx:18]: They shouldn't have too. We're just mapping the existing structure to the new structure.

File: modules/react/button/lib/Button.tsx (hunk 11 above)
Comment (josh-bagwell) [reply modules/react/button/lib/Button.tsx:67]: Yup. This is the mediator for each button variant. Just allows us to clean stuff up.

File: modules/preview-react/pill/lib/Pill.tsx (hunk 12 above)
Comment (josh-bagwell) [reply modules/preview-react/pill/lib/Pill.tsx:141]: Yeah this just got updated. I made `focus` a class and `focus-visible` the psuedo

File: modules/react/button/lib/DeleteButton.tsx [hunk 13]
```diff
@@ -1,75 +1,54 @@
...
-    label: themeError.contrast,
-  },
-  hover: {
-    background: themeError.dark,
+/**
+ * Extends all the style properties from Box to our buttons as well as props from ButtonContainerProps.
+ * We omit `ref` since all of our buttons use `createComponent` and already give access to `ref`.
+ * Use this type to extend and customize any one off buttons that you want full control over styling.
+ */
+export interface DeleteButtonProps extends Omit<ButtonContainerProps, 'ref'> {}
```
[the remark this answers is not in the document]
Comment (josh-bagwell) [reply modules/react/button/lib/DeleteButton.tsx:14]: Ahh. Yeah I left that in since it's an exportable interface and if anyone was using that, I didn't want it to cause an issue if I removed it. That goes for all the other buttons as well.

File: modules/react/button/lib/DeleteButton.tsx [hunk 14]
```diff
@@ -1,75 +1,54 @@
...
+    [buttonVars.disabled.label]: cssVar(brand.error.accent),
+    opacity: 1,
   },
 });
 
-/**
- * Use sparingly for destructive actions that will result in data loss, can’t be undone, or will
- * have significant consequences. They commonly appear in confirmation dialogs as the final
- * confirmation before deleting.
- */
```
Comment (mannycarrera4) [inline modules/react/button/lib/DeleteButton.tsx:48]: I think we want this comment back? above `export const DeleteButton...`

File: modules/react/button/lib/Button.tsx [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        {icon && iconPosition === 'end' && (
+          <BaseButton.Icon
+            size={size}
+            iconPosition={iconPosition}
+            icon={icon}
+            shouldMirrorIcon={shouldMirrorIcon}
+          />
+        )}
+      </BaseButton>
+    );
```
[the remark this answers is not in the document]
Comment (josh-bagwell) [reply modules/react/button/lib/Button.tsx:65]: Yeah definitely. 

It's not accessible via the package since I didn't add it to the index. But I can callout what it's job is.

File: modules/react/button/lib/PrimaryButton.tsx [hunk 16]
```diff
@@ -1,171 +1,104 @@
...
-  }
-};
-
-export interface PrimaryButtonProps extends Themeable, GrowthBehavior, BaseButtonProps {
+/**
+ * Extends all the style properties from Box to our buttons as well as props from ButtonContainerProps.
+ * We omit `ref` since all of our buttons use `createComponent` and already give access to `ref`.
+ * Use this type to extend and customize any one off buttons that you want full control over styling.
+ */
+export interface PrimaryButtonProps extends Omit<ButtonContainerProps, 'ref'> {
```
Comment (mannycarrera4) [inline modules/react/button/lib/PrimaryButton.tsx:14]: Same thing here, we can extend `ButtonProps` from `Button.tsx`

File: modules/react/button/lib/Button.tsx [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+export interface ButtonProps extends Omit<ButtonContainerProps, 'ref'> {}
+
+export const Button = createComponent('button')({
+  displayName: 'Button',
+  Component: (
+    {
+      children,
+      icon,
+      colors,
+      iconPosition = children ? undefined : 'only',
```
Comment (mannycarrera4) [inline modules/react/button/lib/Button.tsx:20]: what happens with the `start | end` positions?

Comment (josh-bagwell) [reply modules/react/button/lib/Button.tsx:20]: This was just fixed. I fixed it when you called out the actionBar issue

File: modules/react/button/lib/DeleteButton.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (josh-bagwell) [reply modules/react/button/lib/DeleteButton.tsx:14]: Ahh I see what you mean

File: modules/react/button/lib/DeleteButton.tsx (hunk 14 above)
Comment (josh-bagwell) [reply modules/react/button/lib/DeleteButton.tsx:48]: Yeah definitely. Re-added.

File: modules/react/button/lib/Button.tsx [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+export interface ButtonProps extends Omit<ButtonContainerProps, 'ref'> {}
+
+export const Button = createComponent('button')({
+  displayName: 'Button',
+  Component: (
+    {
+      children,
+      icon,
+      colors,
+      iconPosition = icon ? (children ? 'start' : 'only') : undefined,
```
Comment (mannycarrera4) [inline modules/react/button/lib/Button.tsx:20]: Does `only` do anything? In the primary button storybook example, iconPosition isn't reference, there's just no children.

Comment (mannycarrera4) [reply modules/react/button/lib/Button.tsx:20]: image

**Comment by mannycarrera4** [comment]
Focus seems to stay visible even after clicking the button.
Repo:
1. Go to primary button storybook example
2. Hit tab to focus the first button
3. Click down on the first button
4. Notice that the focus ring stays visible even after clicking with the mouse

[Link](https://recordit.co/98AmXD4E7R)

File: modules/react/button/lib/Button.tsx (hunk 18 above)
Comment (josh-bagwell) [reply modules/react/button/lib/Button.tsx:20]: Yup. 

`only` refers to when a Button has only an icon with no children. So it maps to largeOnly, mediumOnly, smallOnly and extraSmallOnly

Comment (mannycarrera4) [reply modules/react/button/lib/Button.tsx:20]: Where does that value of the prop be used? In the icon only variant, iconPosition isn't defined

**Comment by mannycarrera4** [comment]
I think we should update the `CustomStyles` storybook example to use `Button` and maybe an example or two using the `cs` prop and `createStyles` to show people how to override things

File: modules/react/button/lib/Button.tsx (hunk 18 above)
Comment (josh-bagwell) [reply modules/react/button/lib/Button.tsx:20]: Which shouldn't cause an issue for users

File: modules/react/button/lib/Button.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (mannycarrera4) [reply modules/react/button/lib/Button.tsx:65]: so no one can consume Button.tsx?

File: modules/react/button/lib/Button.tsx (hunk 18 above)
Comment (mannycarrera4) [reply modules/react/button/lib/Button.tsx:20]: It seems like I can still pass `only` to a button that has children and an icon and it breaks the padding. I'm just curious why we need `only`

File: modules/react/button/lib/DeleteButton.tsx [hunk 19]
```diff
@@ -1,43 +1,44 @@
...
-    background: themeError.main,
-    label: themeError.contrast,
-  },
-  hover: {
-    background: themeError.dark,
+/**
+ * Extends all the style properties from Box to our buttons as well as props from ButtonContainerProps.
+ * We omit `ref` since all of our buttons use `createComponent` and already give access to `ref`.
+ * Use this type to extend and customize any one off buttons that you want full control over styling.
+ */
```
Comment (mannycarrera4) [inline modules/react/button/lib/DeleteButton.tsx:13]: I don't think you need this comment here, since the original prop interface has it. Same for the other buttons

File: modules/react/button/lib/Button.tsx [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+    Element
+  ) => {
+    return (
+      <BaseButton
+        as={Element}
+        ref={ref}
+        size={size}
+        icon={icon}
+        iconPosition={iconPosition}
+        type="button"
```
Comment (mannycarrera4) [inline modules/react/button/lib/Button.tsx:35]:
```suggestion
```

Comment (mannycarrera4) [reply modules/react/button/lib/Button.tsx:35]: this is already added on base button https://github.com/Workday/canvas-kit/pull/2285/files#diff-417ee64d3f922776d0d7fc16914309849f3820955d4110431c675c0f4043d885R35

File: modules/react/button/lib/DeleteButton.tsx [hunk 21]
```diff
@@ -48,28 +49,11 @@ const getDeleteButtonColors = ({
...
-      colors={getDeleteButtonColors(theme)}
-      size={size}
-      {...elemProps}
-    >
-      <BaseButton.Label>{children}</BaseButton.Label>
-    </BaseButton>
-  ),
+  Component: ({children, colors, size, ...elemProps}: DeleteButtonProps, ref, Element) => {
+    return (
+      <Button as={Element} ref={ref} type="button" size={size} cs={[deleteStyles]} {...elemProps}>
```
Comment (mannycarrera4) [inline modules/react/button/lib/DeleteButton.tsx:54]:
```suggestion
      <Button as={Element} ref={ref} size={size} cs={[deleteStyles]} {...elemProps}>
```

File: modules/react/button/lib/DeleteButton.tsx (hunk 19 above)
Comment (josh-bagwell) [reply modules/react/button/lib/DeleteButton.tsx:13]: If this is removed, then DeleteButtonProps won't have JSDocs for it. You'd have to hover over ButtonProps directly. Is that what we want to do ?

File: modules/react/button/lib/Button.tsx (hunk 18 above)
Comment (mannycarrera4) [reply modules/react/button/lib/Button.tsx:20]: Ok i think i know what's going on! this line here in `Button.tsx`:
```
iconPosition = icon ? (children ? 'start' : 'only') : undefined,
```
Is doing some funky stuff. You're changing the value here, but down below you have this logic:
```
 {icon && iconPosition === 'only' && (
          <BaseButton.Icon
            size={size}
            iconPosition={iconPosition}
            icon={icon}
            shouldMirrorIcon={shouldMirrorIcon}
          />
        )}
```
But it currently works because when you deconstruct the props, you set it to undefined if there's no children which trumps the logic below in the render function. So adding `iconPosition="only"` does nothing on an icon only button

**Comment by NicholasBoll** [comment]
> Focus seems to stay visible even after clicking the button. Repo:
> 
> 1. Go to primary button storybook example
> 2. Hit tab to focus the first button
> 3. Click down on the first button
> 4. Notice that the focus ring stays visible even after clicking with the mouse
> 
> [Link](https://recordit.co/98AmXD4E7R)

@mannycarrera4 We're no longer using the `InputProvider` to determine if the focus ring should be visible. That was our own heuristics. We're now using [:focus-visible](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible):

> The :focus-visible pseudo-class applies while an element matches the [:focus](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus) pseudo-class and the UA ([User Agent](https://developer.mozilla.org/en-US/docs/Glossary/User_agent)) determines via heuristics that the focus should be made evident on the element. (Many browsers show a "focus ring" by default in this case.)

File: .storybook/preview.js [hunk 22]
```diff
@@ -5,6 +5,9 @@ import routes from './routes';
 import {CanvasProviderDecorator} from '../utils/storybook';
 import theme from './theme';
 import {defaultCanvasTheme} from '@workday/canvas-kit-react/common';
+import "@workday/canvas-tokens-web/dist/css/base/_variables.css";
+import "@workday/canvas-tokens-web/dist/css/brand/_variables.css";
+import "@workday/canvas-tokens-web/dist/css/system/_variables.css";
```
[the remark this answers is not in the document]
Comment (NicholasBoll) [reply .storybook/preview.js:10]: For now, the exports of Canvas Tokens includes the `dist` folder. I think we wanted to remove it in the future, but we don't control that here.

File: package.json [hunk 23]
```diff
@@ -176,6 +176,7 @@
   "browserslist": "> 5%, last 2 firefox versions, last 2 chrome versions, last 2 safari versions, last 2 edge versions, Firefox ESR, not ie 11",
   "dependencies": {
     "@emotion/react": "^11.7.1",
-    "@emotion/styled": "^11.6.0"
+    "@emotion/styled": "^11.6.0",
```
Comment (mannycarrera4) [inline package.json:179]: this shouldn't be down graded

**Review by mannycarrera4 — APPROVED** [review]

**Comment by mannycarrera4** [comment]
EMOTIONAL DAMAGE

File: modules/react/button/lib/BaseButton.tsx [hunk 24]
```diff
@@ -45,260 +56,267 @@ export interface ButtonContainerProps extends Partial<SystemIconProps>, GrowthBe
...
-          ...baseStyles,
-          ...hoverStyles,
-          ...activeStyles,
-          outline: 'none',
-          boxShadow: 'none',
-          animation: 'none',
-        },
-      }),
-    };
+  '&:disabled, &:active:disabled, &:focus:disabled, &:hover:disabled': {
```
Comment (NicholasBoll) [inline modules/react/button/lib/BaseButton.tsx:168]:
```suggestion
  '&:disabled, &:active:disabled, &:focus-visible:disabled, &:hover:disabled': {
```

**Review by NicholasBoll — APPROVED** [review]
